### PR TITLE
feat(web): augment custom query reactivechart

### PR DESCRIPTION
**Pull request by SavvyShah**
- Pass more information to `customQuery` and `defaultQuery` by passing `value` parameter as an object of required information.
- [Notion card](https://www.notion.so/appbase/Change-value-shape-for-chart-type-071c30cec767415986fefb6d173846dd)
- [Loom preview - 5th Sep](https://www.loom.com/share/398090dbbea64d6497e902ec93d0892b)

### CODE REVIEW

**Review by bietkul — CHANGES_REQUESTED** [review]
@SavvyShah Can you share a loom explaining the code changes?

**Comment by SavvyShah** [comment]
> @SavvyShah Can you share a loom explaining the code changes?

https://www.loom.com/share/d17b0c146c454349ac266f4716b2d53d

File: packages/web/src/components/basic/SelectedFilters.js [hunk 1]
```diff
@@ -82,6 +82,10 @@ class SelectedFilters extends Component {
 			if (value.location) {
 				label = `${value.location} - ${label}`;
 			}
+			// Detect if value is from a chart with chartType as custom
+			if (value && value.mainLabel && value.secondaryLabel) {
+				label = `${value.mainLabel}-${value.secondaryLabel}-${JSON.stringify(value.data)}`;
```
Comment (bietkul) [inline packages/web/src/components/basic/SelectedFilters.js]: @SavvyShah What if only `mainLabel` or `secondaryLabel` is present? Selected filters should handle those cases.
You should handle the `null` values as well as mentioned in Notion. Do not display `null` or `undefined` values.

File: packages/web/src/components/basic/URLParamsProvider.js [hunk 2]
```diff
@@ -201,6 +201,9 @@ class URLParamsProvider extends Component {
 			// TODO: support for NestedList
 			if (value.location) return value;
 			if (value.category) return value;
+			if (value.mainLabel && value.secondaryLabel) {
```
Comment (bietkul) [inline packages/web/src/components/basic/URLParamsProvider.js]: Only check the presence of `mainLabel` and `secondaryLabel` keys in value, if `true` then return `value` as it.

File: packages/web/src/components/chart/ReactiveChart.js [hunk 3]
```diff
@@ -66,6 +67,15 @@ class ReactiveChart extends React.Component {
 		}
 		this.setReact(props, this.internalComponent);
 		this.handleRange = debounce(this.handleRange, 100);
+
+		if (props.chartType === ChartTypes.Custom) {
+			if (typeof props.defaultQuery !== 'function' || typeof props.setOption !== 'function') {
+				throw new Error('defaultQuery and setOption should be function when chartType is custom');
```
Comment (bietkul) [inline packages/web/src/components/chart/ReactiveChart.js]:
```suggestion
				throw new Error('defaultQuery and setOption should be defined when chartType is custom');
```

File: packages/web/src/components/chart/ReactiveChart.js [hunk 4]
```diff
@@ -66,6 +67,15 @@ class ReactiveChart extends React.Component {
...
+			}
+			if (props.useAsFilter && typeof props.customQuery !== 'function') {
+				throw new Error('customQuery should be function when chartType is custom and useAsFilter is true');
```
Comment (bietkul) [inline packages/web/src/components/chart/ReactiveChart.js]:
```suggestion
				throw new Error('customQuery should be defined when chartType is custom and useAsFilter is true');
```

File: packages/web/src/components/chart/ReactiveChart.js [hunk 5]
```diff
@@ -343,8 +360,7 @@ class ReactiveChart extends React.Component {
 
 ReactiveChart.generateQueryOptions = (props, after, value = {}) => {
 	const queryOptions = getQueryOptions(props);
-	const valueArray = Array.isArray(value) ? value : [value];
-	return getAggsQuery(valueArray, queryOptions, props);
+	return getAggsQuery(value, queryOptions, props);
```
Comment (bietkul) [inline packages/web/src/components/chart/ReactiveChart.js:363]: @SavvyShah Only make the above change when `chartType` is `custom`. No need to change the existing code.

**Review by bietkul — CHANGES_REQUESTED** [review]

File: packages/web/src/components/basic/SelectedFilters.js (hunk 1 above)
Comment (SavvyShah) [reply packages/web/src/components/basic/SelectedFilters.js]: So if `mainLabel` is empty then it should be like : `${id}-${secondaryLabel}`?

File: packages/web/src/components/chart/ReactiveChart.js (hunk 5 above)
Comment (SavvyShah) [reply packages/web/src/components/chart/ReactiveChart.js:363]: The change is not particular to the custom chart. It currently affects `value` param for functions like `defaultQuery` and `customQuery` in all the charts. [More details in loom video](https://www.loom.com/share/675162c8c79d434e994f56ac2b97adf6).

File: packages/web/src/components/basic/URLParamsProvider.js [hunk 6]
```diff
@@ -201,6 +201,9 @@ class URLParamsProvider extends Component {
 			// TODO: support for NestedList
 			if (value.location) return value;
 			if (value.category) return value;
+			if (Object.hasOwn(value, 'mainLabel') && Object.hasOwn(value, 'secondaryLabel')) {
```
Comment (bietkul) [inline packages/web/src/components/basic/URLParamsProvider.js]:
```suggestion
			if (Object.hasOwn(value, 'mainLabel') || Object.hasOwn(value, 'secondaryLabel')) {
```

**Review by bietkul — CHANGES_REQUESTED** [review]

**Pull request merged**